expression/util.go

```diff
@@ -1794,3 +1798,57 @@ func ExprsToStringsForDisplay(exprs []Expression) []string {
 	}
 	return strs
 }
+
+// SortAndGenMD5HashForCNFExprs sorts cnf exprs by their hash codes: (A or B or C) and (D or E or F) and (G or H or I)
+// Return true and the hash code if none DNF expressions' length > 1000;
+// Otherwise, return false and 0
+// The differences between this function with CanonicalHashCode is:
+// 1. CanonicalHashCode will treat "a < b" the same with "b >= a", while this function won't
+// 2. This function will treat {"A or B or C", "A or C or B", "B or A or C", "B or C or A", "C or A or B", "C or B or A"} all the same,
+// while CanonicalHashCode won't
+// TODO: treat "a > 3" the same with "a > ?" whose parameter is set 3. Currently, the hash code of constant with ParamMarker
+// contains ParamMarker order only, no value included
+func SortAndGenMD5HashForCNFExprs(sc *stmtctx.StatementContext, conditions []Expression) (bool, uint64) {
+	cnfHashCodes := make([][]byte, 0, len(conditions))
+	for _, arg := range conditions {
+		if sf, ok := arg.(*ScalarFunction); ok && sf.FuncName.L == ast.LogicOr {
+			dnfExprs := FlattenDNFConditions(sf)
+			if len(dnfExprs) > 1000 {
+				// It might consume too much cpu and memory resource to compute hash code
+				logutil.BgLogger().Warn("Expression is too complex to record history stat.",
+					zap.Int("DNFExprLen", len(dnfExprs)))
+				return false, 0
+			}
+			dnfHashCodes := make([][]byte, 0, len(dnfExprs))
+			for _, dnf := range dnfExprs {
+				dnfHashCode := dnf.HashCode(sc)
+				dnfHashCodes = checkHashLengthAndAppend(dnfHashCode, dnfHashCodes)
+			}
+			sort.Slice(dnfHashCodes, func(i, j int) bool { return bytes.Compare(dnfHashCodes[i], dnfHashCodes[j]) < 0 })
+
+			flatDnfHashCode := bytes.Join(dnfHashCodes, nil)
+			cnfHashCodes = append(cnfHashCodes, flatDnfHashCode)
```

Comment:
Why don't need to call `checkHashLengthAndAppend`?

Reply:
Yes, it reminds me that I need to restrict the memory usage more carefully here, I'll update it.